numpy/core/src/umath/matmul.c.src

```diff
@@ -267,19 +267,71 @@ NPY_NO_EXPORT void
 
 /**end repeat**/
 
+
+NPY_NO_EXPORT void
+OBJECT_matmul_inner_noblas(void *_ip1, npy_intp is1_m, npy_intp is1_n,
+                           void *_ip2, npy_intp is2_n, npy_intp is2_p,
+                           void *_op, npy_intp os_m, npy_intp os_p,
+                           npy_intp dm, npy_intp dn, npy_intp dp)                         
+{
+    npy_intp m, n, p;
+    npy_intp ib1_n, ib2_n, ib2_p, ob_p;
+    char *ip1 = (char *)_ip1, *ip2 = (char *)_ip2, *op = (char *)_op;
+
+    ib1_n = is1_n * dn;
+    ib2_n = is2_n * dn;
+    ib2_p = is2_p * dp;
+    ob_p  = os_p * dp;
+
+    PyObject *tmp1, *tmp2 = NULL;
+
+    for (m = 0; m < dm; m++) {
+        for (p = 0; p < dp; p++) {
+            if ( 0 == dn ) return;
+            // The first pass where n == 0 has to be separated from n>0 because
+            // there might not be a "zero" for objects, and even if there is we don't know it.
+            // This also means the degenerate case is not really well-defined.
+
+            // n == 0 case
+            PyObject *tmp = PyNumber_Multiply(*((PyObject **)ip1), *((PyObject **)ip2));
+            ip2 += is2_n;
+            ip1 += is1_n;
+            // n > 0 cases
+            for (n = 1; n < dn; n++) {
+                tmp1 = PyNumber_Multiply(*((PyObject **)ip1), *((PyObject **)ip2));
+                tmp2 = PyNumber_Add(tmp, tmp1);
```

Comment:
You need to check for and propagate `NULL` return values from all the PyNumber functions, unfortunately. I don't remember how we handle this in other object loops.

Reply:
Thanks for the review! I extracted the NULL checks from `OBJECT_dot`. I think it should do the job, but is there any way to trigger them from python to test it?